drivers/hmis_external_apis/app/models/hmis_external_apis/tc_hmis/importers/loaders/custom_services_loader.rb

```diff
@@ -0,0 +1,578 @@
+      create_service_types
+      rows = @reader.rows(filename: filename, header_row_number: 2, field_id_row_number: nil)
+      clobber_records(rows) if clobber
+
+      # services is an instance variable because it holds state that is updated by ar_import, and is needed in create_records
+      @services = create_services(rows)
+      ar_import(service_class, @services.values) # We need to save these first as we need the ids for CDEs
+
+      cdes = create_records(rows)
+      ar_import(cde_class, cdes)
+    end
+
+    private def create_service_types
+      # service types should already exist on production
+      return unless Rails.env.development?
+
+      category = Hmis::Hud::CustomServiceCategory.order(:id).last
+      configs.values.each do |value|
+        Hmis::Hud::CustomServiceType.where(data_source_id: data_source.id).where(name: value.fetch(:service_type)).first_or_create! do |st|
+          st.UserID = system_hud_user.id
+          st.custom_service_category_id = category.id
+        end
+      end
+    end
+
+    private def clobber_records(rows)
+      custom_service_ids = Set.new.tap do |ids|
+        rows.each do |row|
+          row_field_value = row.field_value(TOUCHPOINT_NAME)
+          config = configs[row_field_value]
+          next if config.blank?
+
+          ids << generate_service_id(config, row) # Collect only the services in the file to be clobbered
+        end
+      end
+      services = service_class.where(CustomServiceID: custom_service_ids, data_source: data_source.id)
+
+      rails_service_ids = Set.new.tap do |ids|
+        services.find_each do |service|
+          ids << service.id
+        end
+      end
+
+      services.delete_all
+
+      cdes = cde_class.where(owner_type: service_class.name, owner_id: rails_service_ids)
+      cdes.delete_all
+    end
+
+    private def create_services(rows)
+      log_info('creating services')
+      expected = 0
+      actual = 0
+
+      enrollments = Hmis::Hud::Enrollment.where(data_source_id: data_source.id).index_by(&:enrollment_id)
+      service_type_ids = Hmis::Hud::CustomServiceType.where(data_source_id: data_source.id).pluck(:name, :id).to_h
+
+      result = {}.tap do |services|
+        rows.each do |row|
+          expected += 1
+
+          row_field_value = row.field_value(TOUCHPOINT_NAME)
+          config = configs[row_field_value]
+          if config.blank?
+            log_skipped_row(row, field: TOUCHPOINT_NAME)
+            next
+          end
+
+          response_id = row.field_value(RESPONSE_ID)
+
+          service_type = config[:service_type]
+          service_type_id = service_type_ids[service_type]
+          if service_type_id.blank?
+            log_info("Service type configuration error: can't find #{service_type}!")
+            log_skipped_row(row, field: :service_type)
+            next
+          end
+
+          row_field_value = row_enrollment_id(row)
+          next if row_field_value.blank? # enrollment id is blank
```

Comment:
@gigxz something to flag: I exported everything from 2023 to present (selecting just the touchpoints used in this importer) and it looks like about 50% of the records had a blank enrollment_id

Reply:
@ttoomey thank you - that is good to know - I will flag to them